File: searchA2dMatrix.js
// Time Complexity : O(log n)
// Space Complexity : O(1) - constance
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

const searchA2dMatrix = (arr, target) => {
  let low = 0;
  let high = arr.length * arr[0].length - 1;
  let mid = 0;
  let midIdx;

  while (low <= high) {
    midIdx = Math.floor((low + high) / 2);
    mid =
      arr[Math.floor(midIdx / arr[0].length)][
        Math.floor(midIdx % arr[0].length)
      ];

    if (mid === target) {
      return true;
    }

    if (mid < target) {
      low = midIdx + 1;
    } else {
      high = midIdx - 1;
    }
  }
  return false;
};

console.log(
  searchA2dMatrix(
    [
      [1, 3, 5, 7],
      [10, 11, 16, 20],
      [23, 30, 34, 60],
    ],
    7
  )
);

File: Search a 2D Matrix.cpp
class Solution {
public:
    // Time Complexity : O(logn)
    // Space Complexity :  O(1
    // Did this code successfully run on Leetcode :  O(logn)
    // Any problem you faced while coding this :  No

    /*
    We can consider the 2D array as a flattened 1D array
    Since the array is sorted, we can do binary search
    which will take O(log(n*m)) time
    */

    bool searchMatrix(vector<vector<int>>& mat, int target) {
        int m, n;
        m = mat.size();
        n = mat[0].size();

        int lo = 0, hi = m*n;

        while(lo<hi){
          int mid = lo+(hi-lo)/2;

          int row = mid/n, col = mid%n;

          if(mat[row][col] == target)  return true;
          if(mat[row][col] < target)  lo = mid + 1;
          else  hi = mid;
        }

        //if not found
        return false;
    }
};
